src/Network/DNS/FFI.hs

```diff
@@ -55,12 +55,12 @@ withCResState act
   | otherwise = withMVar resolvLock $ \() -> act nullPtr
 
 withCResInit :: Ptr CResState -> IO a -> IO a
-withCResInit stptr act = do
-   rc1 <- c_res_opt_set_use_dnssec stptr
-   unless (rc1 == 0) $
-       fail "res_init(3) failed"
-   resetErrno
-   act `finally` c_res_close stptr
+withCResInit stptr act = flip finally (c_res_close stptr) $ do
```

Comment:
I am confused now.  Shouldn't `c_res_close` only be called after `c_res_opt_set_use_dnssec` was successful?  
I thought this was the reason not to integrate it directly into `withCResState`.

Reply:
`c_res_close` is actually safe to call inside `withCResState`, but this is only lucky coincidence because we do

```haskell
_ <- c_memset ptr 0 sizeOfResState
```

inside `withCResState` after allocation of the state pointer. In this case (all fields, e.g. pointers, etc., are zeros), `c_rec_close` is effectively no-op and thus safe. But logically, it must be called after succesful `res_ninit()`.

By wrapping the whole `withCResInit` in `finally $ cres_close ptr` I use the same lucky assumption as could be used for `withResCState`, but with `withCResInit` this does not expand too far beyond the place of call to `res_ninit()`.

Reply:
... not exactly no-op, see glibc implementation [here](https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/resolv/res-close.c), but at least `statp->nscount` will be 0 and no `free` happen (inside function `__res_iclose()`. Though I worry about possible call to `__close_nocacnel_nostatus()` (because `statp->_vsocks` is 0) and `__resolv_conf_detach()`

Reply:
You really should structure this using `bracket` and not `finally`. 

I think that what @andreasabel was asking.

If `res_ninit` fails, one should assume there isn't anything to close/ cleanup. Docs don't say either way, especially they don't say that it's safe to call `res_close` on pointer which is has failed to `res_ninit`. One *should not* look into documentation to figure out whether it's safe, as there is no contract (= docs) it won't change.

Reply:
... so probably it's safer to keep to the 2nd commit, though it makes possible to leak memory in environments with asynchronous exceptions

Reply:
In `bracket` initialization is run with asynchronous exception masked. I don't understand how

```haskell
bracket init c_res_close $ \sptr' -> act sptr' where
  init = do
    rc1 <- c_res_opt_set_use_dnssec stptr
    unless (rc1 == 0) $ fail "res_init(3) failed"
    resetErrno
```

would leak.

Reply:
I agree that bracket won't leak, just tested with this:

```haskell
withCResInit :: Ptr CResState -> IO a -> IO a
withCResInit stptr act = bracket
     (do
         rc1 <- c_res_opt_set_use_dnssec stptr
         unless (rc1 == 0) $
             fail "res_init(3) failed"
         return ()
     )
     (const $ c_res_close stptr)
     (const $ do
         resetErrno
         act
     )
```

I'll commit this if you don't mind against style

Reply:
`return ()` was excessive after `unless` but I already committed